Use ggcr library to write images with non-distributable layers

DennisDenuto opened the pull request:
There has been an upstream change made to ggcr that allows configuring `Write` and `MultiWrite` to write layers that have a media type of 'non-distributable'.

This PR removes the previous approach of handling writing non-dist layers, and instead delegates to the library.

Another nice benefits (apart from less code) is potential performance improvement due to de-duping layer logic now also applied to non-dist layers

DennisDenuto committed:
Code refactor

- Use ggcr library to write non-distributable layers

Authored-by: Dennis Leon <leonde@vmware.com>

cppforlife commented on pkg/imgpkg/cmd/copy.go:
```diff
@@ -77,9 +75,12 @@ func (o *CopyOptions) Run() error {
 	prefixedLogger := logger.NewPrefixedWriter("copy | ")
 
-	registry, err := ctlimg.NewRegistry(o.RegistryFlags.AsRegistryOpts(), imagelayers.NewImageLayerWriterCheck(o.IncludeNonDistributableFlag.IncludeNonDistributable))
+	opts := o.RegistryFlags.AsRegistryOpts()
```
opts -> registryOpts to make it more obvious what this is in context of copy cmd.

DennisDenuto replied:
ok done.

cppforlife commented on pkg/imgpkg/cmd/copy.go:
```diff
@@ -77,9 +75,12 @@ func (o *CopyOptions) Run() error {
 
-	registry, err := ctlimg.NewRegistry(o.RegistryFlags.AsRegistryOpts(), imagelayers.NewImageLayerWriterCheck(o.IncludeNonDistributableFlag.IncludeNonDistributable))
+	opts := o.RegistryFlags.AsRegistryOpts()
+	opts.IncludeNonDistributableLayers = o.IncludeNonDistributableFlag.IncludeNonDistributable
```
it would be nice if we remove https://github.com/vmware-tanzu/carvel-imgpkg/blob/develop/pkg/imgpkg/cmd/non_distributable_flags.go dedicated struct and just inline field on CopyOptions since it's just a single flag used in one command. this would make `o.IncludeNonDistributableFlag.IncludeNonDistributable` into `o.IncludeNonDistributable`  which i think reads a bit better

cppforlife replied:
any thoughts on renaming flag from `--include-non-distributable` to `--include-non-distributable-layers`?

DennisDenuto replied:
My thoughts are:
- this is a slightly better name for this flag.
- We have already shipped with this flag. It would be a breaking change. I would want to capture this change in its own issue / separate PR as that brings concerns that are orthogonal to this 'refactor' PR. 
-- I don't have enough concern for this flag rename to create a GH issue for it. My take is if we get feedback from users that this flag is confusing I would prioritize it accordingly. 

I will go ahead and remove the `IncludeNonDistributableFlag` struct and add it to the `CopyOptions` instead.

cppforlife replied:
added https://github.com/vmware-tanzu/carvel-imgpkg/issues/112

cppforlife commented on pkg/imgpkg/cmd/copy.go:
```diff
@@ -7,10 +7,8 @@ import (
 	"os"
 
-	"github.com/google/go-containerregistry/pkg/v1/types"
-	"github.com/k14s/imgpkg/pkg/imgpkg/imagelayers"
```
sounds like we can actually delete imagelayers package as well?

cppforlife replied:
scratch that i remembered that it was being used by TarWriter. i think given its single use location, we should move contents of imagelayers into tar writer.

DennisDenuto replied:
I'll move the files to the tarwriter package. I do think there is value in keeping the logic of whether a layer should be written in its own struct and not be inlined into the tarwriter itself.

cppforlife commented on pkg/imgpkg/image/registry.go:
```diff
@@ -35,9 +33,10 @@ const (
+	CACertPaths                   []string
+	VerifyCerts                   bool
+	Insecure                      bool
+	IncludeNonDistributableLayers bool
```
i would throw in a newline before this since first 3 fields are related to connectivity and this one isnt.

DennisDenuto replied:
ok.

DennisDenuto committed:
Refactor

- Add a new line in registryOpts struct fields to separate 'groups'
- Move ImageLayerWriterFilter into tarwriter package since that is the
only place it is used now
- rename variable to make its intent more obvious
- inline the IncludeNonDistributableFlag into the `CopyOptions`

Authored-by: Dennis Leon <leonde@vmware.com>
